src/dynamo0p3.py

```diff
@@ -3053,6 +3421,83 @@ def operator(self, arg):
+        self._arglist.append(arg.name)
+        # Associated scalar parameters
+        nrow = arg.name + "_nrow"
+        _local_args = [nrow]
+        if arg.function_space_to.orig_name != \
+           arg.function_space_from.orig_name:
+            # If the to- and from-spaces are the same then so are ncol and
+            # nrow so we only pass one of them
+            ncol = arg.name + "_ncol"
+            _local_args.append(ncol)
+        bandwidth = arg.name + "_bandwidth"
+        alpha = arg.name + "_alpha"
+        beta = arg.name + "_beta"
+        gamma_m = arg.name + "_gamma_m"
+        gamma_p = arg.name + "_gamma_p"
+        _local_args += [bandwidth, alpha, beta, gamma_m, gamma_p]
+        self._arglist += _local_args
+
+        intent = arg.intent
+        # Declare the associated scalar arguments before the array because
+        # some of them are used to dimension the latter (and some compilers
+        # get upset if this ordering is not followed)
+        self._parent.add(DeclGen(self._parent, datatype="integer",
+                                 intent="in",
+                                 entity_decls=_local_args))
+        # Declare the array that holds the CMA operator
+        # If this is the first argument in the kernel then keep a
+        # note so that we can put subsequent declarations in the
+        # correct location
+        decl = DeclGen(self._parent, datatype="real", kind="r_def",
+                       dimension=",".join([bandwidth,
+                                           nrow, "ncell_2d"]),
+                       intent=intent, entity_decls=[arg.name])
+        self._parent.add(decl)
+        if self._first_arg:
+            self._first_arg = False
+            self._first_arg_decl = decl
+
+    def banded_dofmap(self, function_space):
+        ''' Declare the banded dofmap required for a CMA operator
+        that maps to/from the specified function space '''
+        from f2pygen import DeclGen
+        ndf = get_fs_ndf_name(function_space)
+        dofmap = get_cbanded_map_name(function_space)
+        self._parent.add(DeclGen(self._parent, datatype="integer",
+                                 dimension=",".join([ndf, "nlayers"]),
+                                 intent="in",
+                                 entity_decls=[dofmap]))
+        self._arglist.append(dofmap)
+
+    def indirection_dofmap(self, function_space, operator=None):
+        ''' Declare the indirection dofmaps required when applying a
+        CMA operator '''
+        from f2pygen import DeclGen
+        if not operator:
+            raise GenerationError("Internal error: no CMA operator supplied.")
+        if operator.type != "gh_columnwise_operator":
+            raise GenerationError(
+                "Internal error: a CMA operator (gh_columnwise_operator) must "
+                "be supplied but got {0}".format(operator.type))
+        # If a kernel applies a CMA operator then it must only have a
```

Comment:
That makes two of us. Must be a hangover from something I had during the development process. I've removed the comment.